### PR TITLE
Avoid getting nAn radius values with computeBoundingSphere in BufferGeometry

### PR DESCRIPTION
**Pull request by tentone**
Hi

I was getting nAn values for when calculating the bounding sphere of the geometry obtained using the PCDLoader (THREE.Points) even when no nAn values were present in the  position attribute.

This PR fixes this and allows cases where there are nAn values in the geometry to still obtain a radius ignore the nAn values.

Thanks a lot

Cumps

### CODE REVIEW

File: src/core/BufferGeometry.js [hunk 1]
```diff
@@ -616,14 +616,14 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
 
 	computeBoundingSphere: function () {
 
-		var box = new Box3();
-		var vector = new Vector3();
-
-		return function computeBoundingSphere() {
+		var box = new THREE.Box3();
```
Comment (Mugen87) [inline src/core/BufferGeometry.js]: Please avoid the usage of the `THREE` prefix in the core.

File: src/core/BufferGeometry.js [hunk 2]
```diff
@@ -636,28 +636,26 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
...
+				var count = array.length;
 
-					vector.x = position.getX( i );
-					vector.y = position.getY( i );
-					vector.z = position.getZ( i );
-					maxRadiusSq = Math.max( maxRadiusSq, center.distanceToSquared( vector ) );
+				for ( var i = 0; i < count; i += 3 ) {
 
-				}
+					vector.set( array[ i ], array[ i + 1 ], array[ i + 2 ] );
```
Comment (Mugen87) [inline src/core/BufferGeometry.js]: We use methods like `.getX()`  so stuff like `InterleavedBufferAttribute` also works. Direct array access breaks this support.

**Review by Mugen87 — CHANGES_REQUESTED** [review]

**Comment by Mugen87** [comment]
In general, i'm not sure to add support for `NaN` values in buffer attributes. I remember we considered this as invalid geometry data.

**Comment by tentone** [comment]
I dont think that we should support NaN values. Loaders should always ensure that all loaded data is valid.

This actually is fixing cases where there is no NaN in the geometry but for some reason the final distance was NaN.

It happens that the change also fixes the calcs for cases where there are NaN values in the geometry.

I have applied the change.

File: src/core/BufferGeometry.js [hunk 3]
```diff
@@ -636,28 +636,25 @@ BufferGeometry.prototype = Object.assign( Object.create( EventDispatcher.prototy
...
+				var count = position.count;
 
-					vector.x = position.getX( i );
-					vector.y = position.getY( i );
-					vector.z = position.getZ( i );
-					maxRadiusSq = Math.max( maxRadiusSq, center.distanceToSquared( vector ) );
+				for ( var i = 0; i < count; i ++ ) {
 
-				}
+					vector.set( position.getX( i ), position.getY( i ), position.getZ( i ) );
```
Comment (Mugen87) [inline src/core/BufferGeometry.js]: This is even more easier:
```js
vector.fromBufferAttribute( position, i );
```

**Comment by Mugen87** [comment]
> This actually is fixing cases where there is no NaN in the geometry but for some reason the final distance was NaN.

Can you show an example when this happens?

**Comment by tentone** [comment]
Yeah, it happens for these 3 files when loaded with PCDLoader by checking the geometry data only one of them actually have a NaN value. The other are fine but the end distance is NaN.

[files.zip](https://github.com/mrdoob/three.js/files/1855860/files.zip)

**Comment by Mugen87** [comment]
I've checked the files and there is actually a bug in `PCDLoader`. In tries to parse an empty line to a point which leads to `NaN` values in the position attribute. I guess we did not notice this because the example uses a binary file. Your files are ASCII based. Can you please insert in this [line](https://github.com/mrdoob/three.js/blob/dev/examples/js/loaders/PCDLoader.js#L169) the following statemtent:
```js
if ( lines[ i ] === '' ) continue;
```
I guess this fix should remove the warning produced in `computeBoundingSphere()`.

**Comment by tentone** [comment]
Yeah, it fixes the problem.

Is there any reason for the bounding sphere to not be calculated from the bounding box?

Since it uses the same center point it should be the same as just using the box center and a radius of half the biggest side of the box.

Thanks a lot

**Comment by Mugen87** [comment]
> Is there any reason for the bounding sphere to not be calculated from the bounding box?

Umm, i don't know. Looking at [Box.getBoundingSphere()](https://github.com/mrdoob/three.js/blob/dev/src/math/Box3.js#L518), i can only imagine an invalid bounding box as the problem source.

**Comment by WestLangley** [comment]
>Is there any reason for the bounding sphere to not be calculated from the bounding box?

That would produce a sphere that is oversized.

>Since it uses the same center point it should be the same as just using the box center and a radius of half the biggest side of the box.

That would miss the corners of the box.

**Comment by tentone** [comment]
> That would produce a sphere that is oversized.
Oh nevermind, you're right, 

I think we can close this PR then?

**Comment by Mugen87** [comment]
Yes. If you like you can add the `vector.fromBufferAttribute( position, i );` thing in a separate PR. It would be a nice improvement of the code.

**Pull request closed without merging**